landa/water_body_management/utils.py

```diff
@@ -0,0 +1,195 @@
+import json
+
+import frappe
+from frappe import _
+from frappe.query_builder.custom import ConstantColumn
+from frappe.utils import cint
+
+
+def create_version_log(doc, event):
+	"""Called via hooks.py to create Version Log of document creation"""
+	if doc.doctype == "File" and doc.attached_to_doctype != "Water Body":
+		return
+
+	doc.flags.updater_reference = {
+		"doctype": doc.doctype,
+		"docname": doc.name,
+		"event": "Created",
+		"label": _("{0} {1} Created").format(doc.doctype, doc.name),
+	}
+
+	doc._doc_before_save = None
+
+
+def get_changed_data(from_datetime: str):
+	"""
+	Get created, modified and deleted Water Body, Fish Species, WBMLO, Attached File data.
+	"""
+	created_modified = get_version_log_query(from_datetime)
+	deleted = get_deleted_document_query(from_datetime)
+	union_query = created_modified.union(deleted)
+	return (frappe.qb.from_(union_query).select(union_query.star).orderby(union_query.creation)).run(
+		as_dict=True
+	)
+
+
+def get_formatted_changes(changes_list: list):
+	"""
+	Format changes dict to be displayed in the Change Log API.
+	"""
+	changes = []
+	for entry in changes_list:
+		changed_data = frappe._dict(json.loads(entry.data))
+		event = get_event(entry, changed_data)
+
+		# Modified dependencies (File and WBMLO)
+		if entry.doctype in ["File", "Water Body Management Local Organization"]:
+			change_log = log_dependency_change(entry, changed_data)
+			if change_log:
+				changes.append(change_log)
+
+			continue
+
+		# Created/Deleted Water Body/Fish Species
+		if event in ["Created", "Deleted"]:
+			changes.append(
+				{
+					"doctype": entry.doctype,
+					"docname": entry.docname,
+					"event": event,
+					"datetime": entry.creation,
+				}
+			)
+			continue
+
+		# Modified Water Body/Fish Species Log
+		change_log = log_modified_change(entry, changed_data, event)
+		if change_log:
+			changes.append(change_log)
+
+	return changes
+
+
+def get_version_log_query(from_datetime: str):
+	version = frappe.qb.DocType("Version")
+	file = frappe.qb.DocType("File")
+	return (
+		frappe.qb.from_(version)
+		.left_join(file)
+		.on(version.docname == file.name)
```

Comment:
```suggestion
		.on(version.ref_doctype == "File" & version.docname == file.name)
```

In case any other doctype is coincidentally named like a file.